### PR TITLE
faiss: 1.8.0->1.9.0

### PR DESCRIPTION
**Pull request by sarahec**
## Description of changes

- Upgraded version 1.8.0 with 1.9.0. 
- Removed passthru testing. This simply ran one of the demo programs. That porogram is no longer maintained and doesn't compile. The newer demos require manual setup and thus aren't suitable.

### Downstream effects

1. Unblocks building on Darwin
2. Supports SWIG.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sarahec** [comment]
Awaiting the outcome of https://github.com/facebookresearch/faiss/issues/3846 before opening for review.

**Comment by emilazy** [comment]
Is this a SWIG 4 thing? We may be able to apply whatever upstream patch fixed it to the stable release.

**Comment by sarahec** [comment]
> Is this a SWIG 4 thing? We may be able to apply whatever upstream patch fixed it to the stable release.

It's more that SWIG 4.2.x, they had to make other changes to their .swig file so it would generate Darwin-buildable code. I tested the faiss nightly yesterday and they're considering cutting a new release (because I asked nicely).

File: pkgs/development/libraries/science/math/faiss/default.nix [hunk 1]
```diff
@@ -54,18 +54,18 @@ stdenv.mkDerivation {
...
   # https://github.com/facebookresearch/faiss/issues/3239
   # that fixes building faiss with swig 4.2.x
-  postPatch = ''
-    substituteInPlace faiss/python/swigfaiss.swig \
-      --replace-fail '#ifdef SWIGWORDSIZE64' '#if (__SIZEOF_LONG__ == 8)'
-  '';
+  # postPatch = ''
+  #   substituteInPlace faiss/python/swigfaiss.swig \
+  #     --replace-fail '#ifdef SWIGWORDSIZE64' '#if (__SIZEOF_LONG__ == 8)'
+  # '';
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/faiss/default.nix:68]: Might as well delete

**Comment by SomeoneSerge** [comment]
@ofborg build faiss python3Packages.faiss

(needs a commit message in the `${attrPath}: ${message}` format for automation)

**Comment by sarahec** [comment]
@SomeoneSerge if faiss comes up with a new release (hopefully), I'll turn this into a pure upgrade, remove the patches, and mark it ready for review. If they don't, I expect to close this.

They have some patches in their history that we might cherry-pick to solve the bug, but I tried the likely candidates and it didn't work out.

**Comment by sarahec** [comment]
They expect to produce a release next week. If and when that happens, we'll update this and enter it for review.

**Comment by sarahec** [comment]
@SomeoneSerge it's ready for review

**Comment by sarahec** [comment]
@ofborg build faiss python3Packages.faiss

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 341418` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
    <li>sqlite-vss</li>
  </ul>
</details>
<details>
  <summary>29 packages built:</summary>
  <ul>
    <li>faiss</li>
    <li>faiss.demos</li>
    <li>faiss.dist</li>
    <li>open-webui</li>
    <li>open-webui.dist</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.colbert-ai</li>
    <li>python311Packages.colbert-ai.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.colbert-ai</li>
    <li>python312Packages.colbert-ai.dist</li>
    <li>python312Packages.faiss</li>
    <li>python312Packages.faiss.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by sarahec** [comment]
Passthru testing is failing due to the demo app it uses being unmaintained (and has been broken since at least 1.8.0). I'm going to remove it.

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 341418` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
    <li>sqlite-vss</li>
  </ul>
</details>
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>faiss</li>
    <li>faiss.dist</li>
    <li>open-webui</li>
    <li>open-webui.dist</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.colbert-ai</li>
    <li>python311Packages.colbert-ai.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.colbert-ai</li>
    <li>python312Packages.colbert-ai.dist</li>
    <li>python312Packages.faiss</li>
    <li>python312Packages.faiss.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by sarahec** [comment]
@ofborg build faiss

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 341418` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
    <li>sqlite-vss</li>
  </ul>
</details>
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>faiss</li>
    <li>faiss.dist</li>
    <li>open-webui</li>
    <li>open-webui.dist</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.colbert-ai</li>
    <li>python311Packages.colbert-ai.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.dist</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.colbert-ai</li>
    <li>python312Packages.colbert-ai.dist</li>
    <li>python312Packages.faiss</li>
    <li>python312Packages.faiss.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
  </ul>
</details>

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 341418` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
    <li>sqlite-vss</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
  </ul>
</details>
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>faiss</li>
    <li>faiss.dist</li>
    <li>faissWithCuda</li>
    <li>faissWithCuda.dist</li>
    <li>open-webui</li>
    <li>open-webui.dist</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.colbert-ai</li>
    <li>python311Packages.colbert-ai.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.dist</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.colbert-ai</li>
    <li>python312Packages.colbert-ai.dist</li>
    <li>python312Packages.faiss</li>
    <li>python312Packages.faiss.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

File: pkgs/development/libraries/science/math/faiss/default.nix [hunk 2]
```diff
@@ -114,32 +94,10 @@ stdenv.mkDerivation {
...
-    + lib.optionalString pythonSupport ''
-      mkdir "$dist"
-      cp faiss/python/dist/*.whl "$dist/"
-    '';
-
-  passthru = {
-    inherit cudaSupport cudaPackages pythonSupport;
-
-    tests = {
-      runDemos =
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/faiss/default.nix:133]: Would be nice to have more thorough integration testing in the long term, but I guess right now this is more of a maintenance burden

**Review by SomeoneSerge — APPROVED** [review]

File: pkgs/development/libraries/science/math/faiss/default.nix (hunk 2 above)
Comment (sarahec) [reply pkgs/development/libraries/science/math/faiss/default.nix:133]: Agreed. The demos don't seem to go through their CI, alas.

File: pkgs/development/libraries/science/math/faiss/default.nix
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/faiss/default.nix:1]: P.S. @sarahec sorry about missing the pings and thanks for the patience

**Comment by trofi** [comment]
The removal of `passthru` in `faiss`:

```
  passthru = {
    inherit cudaSupport cudaPackages pythonSupport;
```

broke pass thru attributes of python bindings:

```
$ nix repl -f.
nix-repl> python3Packages.faiss.passthru
  cudaPackages = «error: attribute 'cudaPackages' missing»;
  cudaSupport = «error: attribute 'cudaSupport' missing»;
  pythonSupport = «error: attribute 'pythonSupport' missing»;
```

Should those be removed from bindings? Or restored in the library?

**Comment by trofi** [comment]
Proposed simple attribute restore as:
- https://github.com/NixOS/nixpkgs/pull/348104

**Pull request merged**